### PR TITLE
Auto increment slugs on per model basis

### LINKED ISSUE #10147: Add ability for slugs to be auto incremented on a per model basis
### Is your proposal related to a problem?

I would like to be able to use auto incrementing page slugs in the case of a duplicate slug within the current parent page's URL path.

Example use case: An event calendar where I don't really care so much about the slugs, but there is a high potential for duplicates with the types of events that may be submitted.

The ability to generate 'nice' URLs is a positive selling point of Wagtail that we don't want to undermine but we do want to help users create pages easily.

### Describe the solution you'd like

`Page` models can have an `auto_increment_slug` meta flag, instead of a global Django setting.

When the value is `True`, and pages are created as children of that page (important: the slug behaviour will be determined by parent models, not the page models itself) will have an extra step on form creation.

If a page is being created (not edited) and there is a value error because of a duplicate URL path, instead of failing the form submission it will automatically append an incremented number to the end of the URL. This incrementing number would need to parse the conflicting page path for the last number and then add to it.

**Edge cases to consider**

* What if the path part is not really meant to be an incremental number (E.g. `path/to/blog-post-about-pi-being-3`)
* Confirmation that we want the flag on parent pages and not the page itself that is being created
* Supporting simple numbers only?
* What if the number is `0`, should we increment from `1`
* Should we provide the ability to set the start number `auto_increment_slug = 0` or `auto_increment_slug = 10` or just always assume it is 1 and the first page that contains the slug without any increment will be the zeroth item (without the `-0`)

**Requirements**

* Unit tests are a must
* Documentation for this feature to be added - maybe in https://docs.wagtail.org/en/latest/advanced_topics/customisation/page_editing_interface.html#customising-generated-forms & also page model reference

### Describe alternatives you've considered

* Do not try to solve this in an automated way and keep the page validation as is, users can decide on what the slug should be manually.

### Additional context

* An older PR has an approach (settings) to this problem - see https://github.com/wagtail/wagtail/pull/2454

> This proposed behaviour has come up again independently in #3477. We've now discussed this among the core team - while auto-rewriting of slugs is a behaviour we'd want to avoid in the default case (because the ability to generate 'nice' URLs is a positive selling point of Wagtail that we don't want to undermine...) we're agreed that this would be a worthwhile feature to support on a per-model basis (perhaps with a `auto_increment_slug` meta flag).
> _Originally posted by @gasman in https://github.com/wagtail/wagtail/issues/2454#issuecomment-290208940_ (March 2017)

### PR DESCRIPTION
**Pull request by NXPY123**
Fixes #10147 Adds ability to auto increment slug on per model basis

-   [x] Do the tests still pass?[^1]
-   [x] Does the code comply with the style guide?
    -   [x] Run `make lint` from the Wagtail root.
-   [ ] For Python changes: Have you added tests to cover the new/fixed behaviour?
-   [x] For new features: Has the documentation been updated accordingly?

### CODE REVIEW

File: wagtail/admin/views/pages/copy.py [hunk 1]
```diff
@@ -22,8 +22,8 @@ def copy(request, page_id):
     # Check if the user has permission to publish subpages on the parent
     can_publish = parent_page.permissions_for_user(request.user).can_publish_subpage()
 
-    # Create the form
-    form = CopyForm(
+    form_class = getattr(page.specific_class, "copy_form_class", CopyForm)
```
Comment (lb-) [inline wagtail/admin/views/pages/copy.py:25]: Just a reminder that there will need to be a test for this override added. 

Maybe even with the code of a custom form that you have in the docs.

Comment (NXPY123) [reply wagtail/admin/views/pages/copy.py:25]: I've added a test. Can you please review it?

**Comment by NXPY123** [comment]
Thanks! I just wanted to say how brilliant the solution you proposed was. It fixed the problem with minimal code change while also being generic enough to cater to other custom form requirements. Extremely elegant!

**Comment by lb-** [comment]
Yeah. I'm pretty pumped for this feature, I hope it will be really useful for 6.0 and beyond. Sometimes it's good to thrash around a few ideas and see what feels right before going all in for a PR.

**Pull request merged**